node/src/proxy_client/stream_handler_pool.rs

```diff
@@ -1218,10 +1230,108 @@ mod tests {
+            let client_request_payload = ClientRequestPayload_0v1 {
+                stream_key,
+                sequenced_packet: SequencedPacket {
+                    data: b"These are the times".to_vec(),
+                    sequence_number: 0,
+                    last_data: false,
+                },
+                target_hostname: Some(String::from("blockedwebsite.com")),
+                target_port: HTTP_PORT,
+                protocol: ProxyProtocol::HTTP,
+                originator_public_key: originator_key,
+            };
+            let package = ExpiredCoresPackage::new(
+                SocketAddr::from_str("1.2.3.4:1234").unwrap(),
+                Some(make_wallet("consuming")),
+                make_meaningless_route(),
+                client_request_payload.into(),
+                0,
+            );
+            let resolver = ResolverWrapperMock::new()
+                .lookup_ip_parameters(&lookup_ip_parameters)
+                .lookup_ip_success(vec![IpAddr::from_str("0.0.0.0").unwrap()]);
+            let proxy_client_sub = peer_actors
+                .proxy_client_opt
+                .clone()
+                .unwrap()
+                .inbound_server_data;
+            let mut subject = StreamHandlerPoolReal::new(
+                Box::new(resolver),
+                cryptde,
+                peer_actors.accountant.report_exit_service_provided.clone(),
+                peer_actors.proxy_client_opt.clone().unwrap(),
+                100,
+                200,
+            );
+            let (stream_killer_tx, stream_killer_rx) = unbounded();
+            subject.stream_killer_rx = stream_killer_rx;
+            let inner_clone = Arc::clone(&subject.inner);
+            let handle = thread::spawn(move || {
+                let mut inner = inner_clone.lock().unwrap();
+                inner.logger = Logger::new(test_name);
+            });
+            handle.join().unwrap();
+            let (stream_adder_tx, _stream_adder_rx) = unbounded();
+            let establisher = StreamEstablisher {
+                cryptde,
+                stream_adder_tx,
+                stream_killer_tx,
+                stream_connector: Box::new(
+                    StreamConnectorMock::new()
+                        .connect_pair_result(Err(Error::from(ErrorKind::Other))),
+                ),
+                proxy_client_sub,
+                logger: subject.inner.lock().unwrap().logger.clone(),
+                channel_factory: Box::new(FuturesChannelFactoryReal {}),
+            };
+
+            subject.inner.lock().unwrap().establisher_factory =
+                Box::new(StreamEstablisherFactoryMock {
+                    make_results: RefCell::new(vec![establisher]),
+                });
+
+            run_process_package_in_actix(subject, package);
+        });
+
+        proxy_client_awaiter.await_message_count(2);
+        let proxy_client_recording = proxy_client_recording_arc.lock().unwrap();
+        assert_eq!(
```

Comment:
As a general philosophical concept, when the originating Node gets a `DnsResolveFailure` message back, it ought to free the same stream resources that are freed by an `InboundServerData` with a `last_data` of `true`. This seems so clear to me that I think we should have caught it in the original iteration from Substratum. Can you look to see what happens when the originating Node gets a DNS failure? If it doesn't free the resources, we should make a card about that. Sending the extra `InboundServerData` package will cost the user needlessly.

Reply:
We have found that the originating Node is reusing the stream key for DNS retries.
Due to this if originating node receives an `ExpiredCoresPackage` with `DnsResolveFailure` it wont remove the stream yet.